src/edu/washington/escience/myria/operator/agg/Aggregate.java

```diff
@@ -138,20 +160,27 @@ protected void init(final ImmutableMap<String, Object> execEnvVars) throws DbExc
     Schema inputSchema = getChild().getSchema();
     Preconditions.checkState(inputSchema != null, "unable to determine schema in init");
     internalAggs = new ArrayList<Aggregator>();
-    emitAggs = new ArrayList<Aggregator>();
+    emitEvals = new ArrayList<GenericEvaluator>();
+    Schema groupingSchema = inputSchema.getSubSchema(gfields);
     Schema stateSchema = Schema.EMPTY_SCHEMA;
     for (int i = 0; i < factories.length; ++i) {
-      internalAggs.addAll(
-          factories[i]
-              .generateInternalAggs(inputSchema, gfields.length + stateSchema.numColumns()));
-      emitAggs.addAll(
-          factories[i].generateEmitAggs(inputSchema, gfields.length + stateSchema.numColumns()));
-      stateSchema = Schema.merge(stateSchema, factories[i].generateStateSchema(inputSchema));
+      internalAggs.addAll(factories[i].generateInternalAggs(inputSchema));
+      List<Expression> emits = factories[i].generateEmitExpressions(inputSchema);
+      Schema newStateSchema = factories[i].generateStateSchema(inputSchema);
+      stateSchema = Schema.merge(stateSchema, newStateSchema);
+      for (Expression exp : emits) {
+        GenericEvaluator evaluator =
+            new GenericEvaluator(
+                exp, new ExpressionOperatorParameter(newStateSchema, newStateSchema));
+        // if (evaluator.needsCompiling()) {
```

Comment:
I removed everything about `needsCompiling()` in the new commit. First, it doesn't hurt to always compile since it only happens in `init()`, and second, now we need to evaluate tuple-by-tuple when the input and the output tuple batches have different sizes, so we need the compiled expression. Found some old discussions in https://github.com/uwescience/myria/pull/548.